### PR TITLE
Hanging query follow up

### PR DESCRIPTION
**Pull request by MarkDuckworth**
AIs for hanging query issue.

- Update WebChannel to include the fix.
- Integration tests
- Enable useFetchStreams
- Add hardAssert to detect the hung state.

Implements b/308456881 and b/308456686
Fixes https://github.com/firebase/firebase-js-sdk/issues/6118

### CODE REVIEW

**Review by tom-andersen — APPROVED** [review]

File: packages/firestore/src/remote/watch_change.ts [hunk 1]
```diff
@@ -240,6 +240,12 @@ class TargetState {
 
   recordTargetResponse(): void {
     this.pendingResponses -= 1;
+    hardAssert(
+      this.pendingResponses >= 0,
+      'Unexpected state, `pendingResponses` is less than 0. This ' +
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply packages/firestore/src/remote/watch_change.ts]: Thanks you @dconeybe. Both good suggestions. Both implemented.

**Comment by dconeybe** [comment]
@MarkDuckworth Is there a GitHub Issue or Buganizer ticket that is canonically tracking the bug that this PR fixes?

**Review by DellaBitta — APPROVED** [review]

**Comment by tgangso** [comment]
Any ETA on release? Thanks

**Comment by tgangso** [comment]
I noticed the comment in #6118 mentioning this does not fix the issue completely. Commenting is closed in that case, so I will add a comment here. Been stuck on firebase v8 for very long now because of this issue, and just want to comment that version 8 does not have this issue, if that helps.

**Comment by thomasdao** [comment]
I have tested the latest version 10.7.0 with this fix and unfortunately the issue #7652 still happens and the query still hang. @MarkDuckworth

**Comment by MarkDuckworth** [comment]
@tgangso, There was a false negative when validating the fix for #6118 in 10.7.0. That lead to the comment you are referring to. After additional review, the testing shows that the memory issue is fixed in 10.7.0. See the updated comment https://github.com/firebase/firebase-js-sdk/issues/6118#issuecomment-1831996970

**Comment by thomasdao** [comment]
Hi @MarkDuckworth, I thought you have cloned the data so I disabled the access to my Firestore database. I have re-enabled the access, please try again. When I tried with the new version 10.7.0, I saw the error below in the console, and the query still hangs. The issue #7652 on Github has been locked and I can't reopen from my side.

```
@firebase/firestore: Firestore (10.7.0): WebChannelConnection RPC 'Listen' stream 0x5b9a037f transport errored: Wn {type: 'c', target: Hn, g: Hn, defaultPrevented: false, status: 1}
```

**Comment by MarkDuckworth** [comment]
@thomasdao,

Thanks for re-enabling access for your test project. I ran the code today and saw the error message you reported followed by a message with the error code `unavailable`.

```
203.js:2 [2023-12-07T21:43:15.567Z]  @firebase/firestore: Firestore (10.7.0): WebChannelConnection RPC 'Listen' stream 0x6eb9c2c0 transport errored: Wn {type: 'c', target: Hn, g: Hn, defaultPrevented: false, status: 1}
203.js:2 [2023-12-07T21:43:15.569Z]  @firebase/firestore: Firestore (10.7.0): PersistentStream close with error: FirebaseError: [code=unavailable]: The operation could not be completed
```

Code `unavailable` is [a GRPC status from the backend](https://grpc.github.io/grpc/core/md_doc_statuscodes.html) that can be retried. The times I saw this error in your repro the SDK did retry and results were received after a delay. Are you seeing the same?

**Comment by thomasdao** [comment]
Hi @MarkDuckworth, I run the test project again and still see the error `WebChannelConnection RPC 'Listen' stream 0x6eb9c2c0 transport errored`.

I did not see the error `PersistentStream close with error: FirebaseError: [code=unavailable]`. I've waited for more than 10 minutes and the query still can't complete. 

When I changed to the older version of Firebase (10.6.0), the query completes quickly.

**Pull request merged**